src/Components/Components/src/Routing/UrlValueConstraint.cs

```diff
@@ -0,0 +1,176 @@
+            => DateTime.TryParse(str, CultureInfo.InvariantCulture, DateTimeStyles.None, out result);
+
+        private static bool TryParse(ReadOnlySpan<char> str, out decimal result)
+            => decimal.TryParse(str, NumberStyles.Number, CultureInfo.InvariantCulture, out result);
+
+        private static bool TryParse(ReadOnlySpan<char> str, out double result)
+            => double.TryParse(str, NumberStyles.Number, CultureInfo.InvariantCulture, out result);
+
+        private static bool TryParse(ReadOnlySpan<char> str, out float result)
+            => float.TryParse(str, NumberStyles.Number, CultureInfo.InvariantCulture, out result);
+
+        private static bool TryParse(ReadOnlySpan<char> str, out int result)
+            => int.TryParse(str, NumberStyles.Integer, CultureInfo.InvariantCulture, out result);
+
+        private static bool TryParse(ReadOnlySpan<char> str, out long result)
+            => long.TryParse(str, NumberStyles.Integer, CultureInfo.InvariantCulture, out result);
+
+        private static UrlValueConstraint? Create(Type targetType) => targetType switch
+        {
+            var x when x == typeof(string) => new TypedUrlValueConstraint<string>(TryParse),
+            var x when x == typeof(bool) => new TypedUrlValueConstraint<bool>(bool.TryParse),
+            var x when x == typeof(bool?) => new NullableTypedUrlValueConstraint<bool>(bool.TryParse),
+            var x when x == typeof(DateTime) => new TypedUrlValueConstraint<DateTime>(TryParse),
+            var x when x == typeof(DateTime?) => new NullableTypedUrlValueConstraint<DateTime>(TryParse),
+            var x when x == typeof(decimal) => new TypedUrlValueConstraint<decimal>(TryParse),
+            var x when x == typeof(decimal?) => new NullableTypedUrlValueConstraint<decimal>(TryParse),
+            var x when x == typeof(double) => new TypedUrlValueConstraint<double>(TryParse),
+            var x when x == typeof(double?) => new NullableTypedUrlValueConstraint<double>(TryParse),
+            var x when x == typeof(float) => new TypedUrlValueConstraint<float>(TryParse),
+            var x when x == typeof(float?) => new NullableTypedUrlValueConstraint<float>(TryParse),
+            var x when x == typeof(Guid) => new TypedUrlValueConstraint<Guid>(Guid.TryParse),
+            var x when x == typeof(Guid?) => new NullableTypedUrlValueConstraint<Guid>(Guid.TryParse),
+            var x when x == typeof(int) => new TypedUrlValueConstraint<int>(TryParse),
+            var x when x == typeof(int?) => new NullableTypedUrlValueConstraint<int>(TryParse),
+            var x when x == typeof(long) => new TypedUrlValueConstraint<long>(TryParse),
+            var x when x == typeof(long?) => new NullableTypedUrlValueConstraint<long>(TryParse),
+            var x => null
+        };
+
+        public abstract bool TryParse(ReadOnlySpan<char> value, [MaybeNullWhen(false)] out object result);
```

Comment:
This should probably be 
```suggestion
        public abstract bool TryParse(ReadOnlySpan<char> value, out object? result);
```

Reply:
Could you clarify why? Did you spot a case where this would be `null` even when the method returns `true`? Or is there some other reason?

Reply:
You're right. I was under the impression that it's more common to say `NotNullWhen(true)` (i.e. when the value is being used) but the runtime has a mixed bag for these attributes. We'll leave it be